lisa/tests/staging/schedutil.py

```diff
@@ -0,0 +1,271 @@
+        # since the kernel is using unsigned arithmetic, but still check in
+        # case there are some dataframe handling issues)
+        assert not (df['expected_cost_margin'] < 0).any()
+        assert not (df['cost_margin'] < 0).any()
+
+        # "rect" method is accurate here since the signal is really following
+        # "post" steps
+        expected_boost_nrg = series_mean(df['expected_cost_margin'])
+        actual_boost_nrg = series_mean(df['cost_margin'])
+
+        # Check that the total amount of boost is close to expectations
+        lower = max(0, expected_boost_nrg - nrg_threshold_pct)
+        higher = expected_boost_nrg
+        passed_overhead = lower <= actual_boost_nrg <= higher
+
+        # Check the shape of the signal: actual boost must be lower or equal
+        # than the expected one.
+        good_shape_nr = (df['cost_margin'] <= df['expected_cost_margin']).sum()
+
+        df_len = len(df)
+        bad_shape_nr = df_len - good_shape_nr
+        bad_shape_pct = bad_shape_nr / df_len * 100
+
+        # Tolerate a few bad samples that added too much boost
+        passed_shape = bad_shape_pct < bad_samples_threshold_pct
+
+        passed = passed_overhead and passed_shape
+        res = ResultBundle.from_bool(passed)
+        res.add_metric('expected boost energy overhead', expected_boost_nrg, '%')
+        res.add_metric('boost energy overhead', actual_boost_nrg, '%')
+        res.add_metric('bad boost samples', bad_shape_pct, '%')
+
+        # Add some slack metrics and plots
+        pa = PerfAnalysis.from_dir(self.res_dir)
+        for task in pa.tasks:
+            pa.plot_slack_histogram(task)
+            pa.plot_perf_index_histogram(task)
+            pa.plot_latency(task)
+
+        res.add_metric('avg slack', self.get_avg_slack(), 'us')
+        res.add_metric('avg negative slack', self.get_avg_slack(only_negative=True), 'us')
+
+        return res
+
+
+class LargeStepUp(RampBoostTestBase):
+    """
+    A single task whose utilization rises extremely quickly
+    """
+    task_name = "step_up"
+
+    def __init__(self, res_dir, plat_info, cpu, nr_steps):
+        super().__init__(res_dir, plat_info, cpu=cpu)
+        self.nr_steps = nr_steps
+
+    @property
+    def rtapp_profile(self):
+        return self.get_rtapp_profile(self.plat_info, self.cpu, self.nr_steps)
+
+    @classmethod
+    def _from_target(cls, target:Target, *, res_dir:ArtifactPath=None, ftrace_coll:FtraceCollector=None, cpu=None, nr_steps=1) -> 'LargeStepUp':
+        plat_info = target.plat_info
+
+        # Use a big CPU by default to allow maximum range of utilization
+        cpu = cpu if cpu is not None else plat_info["capacity-classes"][-1][0]
+
+        rtapp_profile = cls.get_rtapp_profile(plat_info, cpu, nr_steps)
+
+        # Ensure accurate duty cycle and idle state misprediction on some
+        # boards. This helps having predictable execution.
+        with target.disable_idle_states():
```

Comment:
We keep carrying over an explanation as to why we do this... Perhaps we should stick it in a load-tracking specific "I care about signals so give me a white room` context manager.

Reply:
Good cleanup idea